### PR TITLE
[CopyFilesOverSSH]. Fixed bug with creating target folder.

### PR DESCRIPTION
**Pull request by sdobrodeev**
**Task name**: CopyFilesOverSSH
**Description**: Fixed issue with creating target folder on remote server

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** https://github.com/microsoft/azure-pipelines-tasks/issues/13260 and https://github.com/microsoft/azure-pipelines-tasks/issues/13259

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/CopyFilesOverSSHV0/sshhelper.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import Q = require('q');
 import tl = require('azure-pipelines-task-lib/task');
+var path = require('path');
```
Comment (anton-erofeev) [inline Tasks/CopyFilesOverSSHV0/sshhelper.ts:3]:
```suggestion
const path = require('path');
```
Can we replace it with 'const'?

File: Tasks/CopyFilesOverSSHV0/sshhelper.ts [hunk 2]
```diff
@@ -102,6 +103,12 @@ export class SshHelper {
         if (!this.sftpClient) {
             defer.reject(tl.loc('ConnectionNotSetup'));
         }
+
+        const remotePath = path.dirname(dest);
+        if (!await this.sftpClient.exists(remotePath)) {
+            await this.sftpClient.mkdir(remotePath, true);
```
Comment (egor-bryzgalov) [inline Tasks/CopyFilesOverSSHV0/sshhelper.ts:109]: Thank you! Does it make sense to move it to try-catch block?

**Comment by leantk** [comment]
@sdobrodeev what testing have we done? Did we try Windows and Linux OS's?

**Comment by sdobrodeev** [comment]
> @sdobrodeev what testing have we done? Did we try Windows and Linux OS's?

We manually tested the task on Windows and Linux agents. It works correctly.

File: Tasks/CopyFilesOverSSHV0/sshhelper.ts [hunk 3]
```diff
@@ -97,11 +98,26 @@ export class SshHelper {
...
+        
         var defer = Q.defer<string>();
         if (!this.sftpClient) {
             defer.reject(tl.loc('ConnectionNotSetup'));
         }
+
+        const remotePath = path.dirname(dest);
+        try {
+            if (!await this.sftpClient.exists(remotePath)) {
+                await this.sftpClient.mkdir(remotePath, true);
```
Comment (zachariahcox) [inline Tasks/CopyFilesOverSSHV0/sshhelper.ts:115]: what does true mean here? is this the recursive create option? if so, please always prefer the `parameterName: parameterValue` syntax.

**Comment by junketer** [comment]
Hi, 
I might be wrong, but my pipelines have started failing as it seems like sub directories aren't being copied using Copy files over SSH task

Looks like my root directory is created, but the sub directories aren't when we copy over from our CI drop. i'm sure this was all ok before (Friday was last successful run from what I can see)

So the drop has a set of directories in it, and the task seems like it used to copy all that over, but now doesn't
Could the changes here have changed that behaviour?

Our task specifies '**' in the Content field for all files. Is that still a valid wild card? Documentation seems to indicate that is valid

Getting an error like this 

2020-07-13T20:26:34.4376239Z ##[error]Failed to copy C:\vstsagent\d1\_work\r13\a\_Connection-Builder-CI\drop\CB\product\Connection-builder-full-merge.jar. Failed to upload C:\vstsagent\d1\_work\r13\a\_Connection-Builder-CI\drop\CB\product\Connection-builder-full-merge.jar to /oaglocal/dev/connectionbuilderpoc/regression/20200713-merge-3/product\Connection-builder-full-merge.jar on remote machine. Error: put->put: Bad path: /oaglocal/dev/connectionbuilderpoc/regression/20200713-merge-3/product parent not exist.

I was cleaning this location up at the end, but disabling that,  I end up with the root directory

/oaglocal/dev/connectionbuilderpoc/regression/20200713-merge-3 as empty

The drop file root we are copying from has the product directory, as you can see in the source. Drop root source is 

$(System.DefaultWorkingDirectory)/_Connection-Builder-CI/drop/CB

Any help welcome.
Regards

**Comment by leantk** [comment]
@junketer we reverted the change in another PR. Can you verify if you are unblocked now?

**Comment by bjs169** [comment]
> @junketer we reverted the change in another PR. Can you verify if you are unblocked now?

It is working now. Successful deploy. All weekend it wasn't working. Last time I tried about six hours ago it was not working. So looks like whatever you did fixed it.

**Comment by travispalmer** [comment]
> > @junketer we reverted the change in another PR. Can you verify if you are unblocked now?
> 
> It is working now. Successful deploy. All weekend it wasn't working. Last time I tried about six hours ago it was not working. So looks like whatever you did fixed it.

My pipelines are back to working as well. Thanks!

**Comment by junketer** [comment]
> > @junketer we reverted the change in another PR. Can you verify if you are unblocked now?
> 
> It is working now. Successful deploy. All weekend it wasn't working. Last time I tried about six hours ago it was not working. So looks like whatever you did fixed it.

Hi, yes looks like my pipelines are running again thank you
If it helps, my pipelines are copying from an agent Windows box to a Linux target
Thanks for sorting!

**Pull request closed without merging**